types/src/traits/signature_key/bn254/bn254_pub.rs

```diff
@@ -0,0 +1,168 @@
+use super::{BN254Priv, EncodedPublicKey, EncodedSignature, SignatureKey};
+use bincode::Options;
+use bitvec::prelude::*;
+use blake3::traits::digest::generic_array::GenericArray;
+use ethereum_types::U256;
+use hotshot_primitives::qc::bit_vector::{
+    BitVectorQC, QCParams as JFQCParams, StakeTableEntry as JFStakeTableEntry,
+};
+use hotshot_primitives::qc::QuorumCertificate;
+use hotshot_utils::bincode::bincode_opts;
+use jf_primitives::signatures::bls_over_bn254::{BLSOverBN254CurveSignatureScheme, VerKey};
+use jf_primitives::signatures::SignatureScheme;
+use serde::{Deserialize, Serialize};
+use std::{cmp::Ordering, fmt::Debug};
+use tracing::{debug, instrument, warn};
+use typenum::U32;
+
+/// Public key type for an bn254 [`SignatureKey`] pair
+///
+/// This type makes use of noise for non-determinisitc signatures.
+#[derive(Clone, PartialEq, Eq, Hash, Copy, Serialize, Deserialize, Debug)]
+
+pub struct BN254Pub {
+    /// The public key for this keypair
+    pub_key: VerKey,
+}
+
+impl PartialOrd for BN254Pub {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        let self_bytes = &self.pub_key.to_string();
+        let other_bytes = &other.pub_key.to_string();
+        self_bytes.partial_cmp(other_bytes)
+    }
+}
+
+impl Ord for BN254Pub {
+    fn cmp(&self, other: &Self) -> Ordering {
+        let self_bytes = &self.pub_key.to_string();
+        let other_bytes = &other.pub_key.to_string();
+        self_bytes.cmp(other_bytes)
+    }
+}
+
+impl SignatureKey for BN254Pub {
+    type PrivateKey = BN254Priv;
+    type StakeTableEntry = JFStakeTableEntry<VerKey>;
+    type QCParams = JFQCParams<
+        <BLSOverBN254CurveSignatureScheme as SignatureScheme>::VerificationKey,
+        <BLSOverBN254CurveSignatureScheme as SignatureScheme>::PublicParameter,
+    >;
+    type QCType = (
+        <BLSOverBN254CurveSignatureScheme as SignatureScheme>::Signature,
+        BitVec,
+    );
+    // <BitVectorQC<BLSOverBN254CurveSignatureScheme> as AssembledQuorumCertificate<BLSOverBN254CurveSignatureScheme>>::QC;
+
+    #[instrument(skip(self))]
+    fn validate(&self, signature: &EncodedSignature, data: &[u8]) -> bool {
+        let ver_key = self.pub_key;
+        let x: Result<<BLSOverBN254CurveSignatureScheme as SignatureScheme>::Signature, _> =
+            bincode_opts().deserialize(&signature.0);
+        match x {
+            Ok(s) => {
+                // This is the validation for QC partial signature before append().
+                let generic_msg: &GenericArray<u8, U32> = GenericArray::from_slice(data);
+                BLSOverBN254CurveSignatureScheme::verify(&(), &ver_key, generic_msg, &s).is_ok()
+            }
+            Err(_) => false,
+        }
+    }
+
+    fn sign(sk: &Self::PrivateKey, data: &[u8]) -> EncodedSignature {
+        let generic_msg = GenericArray::from_slice(data);
+        let agg_signature_test_wrap = BitVectorQC::<BLSOverBN254CurveSignatureScheme>::sign(
```

Comment:
Why is the word "test" mentioned several time in this function?

Reply:
Fixed this in the latest commit!